### PR TITLE
Feature to report external impacts

### PR DESCRIPTION
**Pull request by dennis-behm**
This enhancement enables the build framework to query for a subset of changed files (see `reportExternalImpactsAnalysisFileFilter`) a set of external collections (see `reportExternalImpactsCollectionPatterns`)  with a given granularity and depth (see `reportExternalImpactsAnalysisDepths`) to generate reports to document cross application impacts.

In this sample, the MortgageApplication was split into two applications App-EPSC and App-EPSM.

App-EPSM provides a shared copybook `epsmtcom.cpy`, which is leveraged by App-EPSC. This enhancement will report this dependency, but will not force App-EPSC to rebuild the program which includes the shared copybook.

The build of App-EPSM contains this reporting (with `--verbose`):

```
** Analyse and report external impacted files.
*** Identified external impacted files for changed file App-EPSM/copybooks-public/epsmtcom.cpy
*** Writing report of external impacts to file /var/jenkins/workspace/App-EPSM-S4/outputs/build.20210722.032455.024/externalImpacts_App-EPSC-master.txt
```

The report (here `externalImpacts_App-EPSC-master.txt`) is written to `workdir` and will contain three columns (logicalFile, filePath and collectionName) with the external impacted files: 

```
EPSCMORT 	 App-EPSC/cobol/epscmort.cbl 	 App-EPSC-master
```

**Important considerations**
- Use the pattern configuration to limit the query to those collections which are relevant
- Running in deep impact analysis mode is an expensive calculation and does recursively identifies impacted files. Please check if this is necessary for you.
- In many cases it should be sufficient to run with the **simple** mode, which performs much faster, because it does not resolve recursively.

### CODE REVIEW

File: utilities/ImpactUtilities.groovy [hunk 1]
```diff
@@ -298,6 +308,110 @@ def scanOnlyStaticDependencies(List buildList, RepositoryClient repositoryClient
...
+
+				def ldepFile = new LogicalDependency(memberName, null, null);
+				repositoryClient.getAllCollections().each{ collection ->
+					String cName = collection.getName()
+					if(matchesPattern(cName,collectionMatcherPatterns)){ // find matching collection names
+						if (cName != props.applicationCollectionName && cName != props.applicationOutputsCollectionName){
+							def Set<String> externalImpactList = collectionImpactsSetMap.get(cName) ?: new HashSet<String>()
+							def logicalFiles = repositoryClient.getAllLogicalFiles(cName, ldepFile);
+							logicalFiles.each{ logicalFile ->
+								def impactRecord = "${logicalFile.getLname()} \t ${logicalFile.getFile()} \t ${cName}"
```
Comment (jbyibm) [inline utilities/ImpactUtilities.groovy:341]: Depending on the length of each item using tab may result in a file difficult to read in text mode. But I don't know what to use. Why not produce a json file ?

File: utilities/ImpactUtilities.groovy [hunk 2]
```diff
@@ -298,6 +308,110 @@ def scanOnlyStaticDependencies(List buildList, RepositoryClient repositoryClient
...
+					def Set<String> externalImpactList = collectionImpactsSetMap.get(externalImpact.getCollection()) ?: new HashSet<String>()
+					def impactRecord = "${externalImpact.getLname()} \t ${externalImpact.getFile()} \t ${externalImpact.getCollection()}"
+					// if (props.verbose) println("*** $impactRecord")
+					externalImpactList.add(impactRecord)
+					collectionImpactsSetMap.put(externalImpact.getCollection(), externalImpactList) // <collection,list of impacted files>
+				}
+			}
+			else {
+				println("*! build property reportExternalImpactsAnalysisDepths has in invalid value : ${props.reportExternalImpactsAnaylsisDepths} , valid: simple | deep")
+			}
```
Comment (jbyibm) [inline utilities/ImpactUtilities.groovy:388]: I think that this control should be in build.groovy during parameters validation.

**Review by jbyibm — CHANGES_REQUESTED** [review]
Just some recommendations/typos

File: build-conf/README.md [hunk 3]
```diff
@@ -50,6 +50,14 @@ dbb.RepositoryClient.userId | DBB configuration property for web application log
 dbb.RepositoryClient.password | DBB configuration property for web application logon password.  ***Can be overridden by build.groovy option -pw, --pw***
 dbb.RepositoryClient.passwordFile | DBB configuration property for web application logon password file.  ***Can be overridden by build.groovy option -pf, --pf***
 
+### dependencyReport.properties
+Propertes used by the impact utilities to generate a report of external impacted files
+
```
[the remark this answers is not in the document]
Comment (dennis-behm) [reply build-conf/README.md]: Good catch 👍

File: utilities/ImpactUtilities.groovy (hunk 1 above)
Comment (dennis-behm) [reply utilities/ImpactUtilities.groovy:341]: Idea of using a text file is to use this as an build file within the external repositories

File: utilities/ImpactUtilities.groovy (hunk 2 above)
Comment (dennis-behm) [reply utilities/ImpactUtilities.groovy:388]: Added an assertion

**Comment by dennis-behm** [comment]
@jbyibm Can you please have a second look at the PR?

**Review by jbyibm — APPROVED** [review]
look good

**Pull request merged**